### PR TITLE
better i/o timeout error on pull

### PR DESCRIPTION
**Pull request by jessfraz**
see: https://forums.docker.com/t/pulling-docker-images-i-o-timeout/740 it is the 2nd most searched for topic on the forum... so I think we should make it less vague.

### CODE REVIEW

File: registry/session.go [hunk 1]
```diff
@@ -404,6 +404,13 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
...
 	res, err := r.client.Do(req)
 	if err != nil {
+		// check if the error is because of i/o timeout
+		// and return a non-obtuse error message for users
+		// "Get https://index.docker.io/v1/repositories/library/busybox/images: i/o timeout"
+		// was a top search on the docker user forum
+		if err.Timeout() {
+			return nil, fmt.Errorf("Network timed out while trying to connect to %s. You may want to check your internet connection or if you are behind a proxy.", repositoryTarget)
+		}
 		return nil, err
```
Comment (cpuguy83) [inline registry/session.go]: While you are in here, this should probably also be something like `fmt.Errorf(error while pulling image: %v", err)`

File: registry/session.go [hunk 2]
```diff
@@ -404,6 +404,13 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
 	req.Header.Set("X-Docker-Token", "true")
 	res, err := r.client.Do(req)
 	if err != nil {
+		// check if the error is because of i/o timeout
+		// and return a non-obtuse error message for users
+		// "Get https://index.docker.io/v1/repositories/library/busybox/images: i/o timeout"
+		// was a top search on the docker user forum
+		if err.Timeout() {
```
Comment (cpuguy83) [inline registry/session.go]: I don't think this does what you think it does (also, CI fail here).

Comment (jfrazelle) [inline registry/session.go]: yeah I thought we were using net package

Comment (jfrazelle) [inline registry/session.go]: now its not pretty

File: registry/session.go [hunk 3]
```diff
@@ -404,7 +404,14 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
 	req.Header.Set("X-Docker-Token", "true")
 	res, err := r.client.Do(req)
 	if err != nil {
-		return nil, err
+		// check if the error is because of i/o timeout
+		// and return a non-obtuse error message for users
+		// "Get https://index.docker.io/v1/repositories/library/busybox/images: i/o timeout"
+		// was a top search on the docker user forum
+		if strings.HasSuffix("i/o timeout", err.Error()) {
```
Comment (cpuguy83) [inline registry/session.go]: This one is backwards :smiling_imp:

Comment (jfrazelle) [inline registry/session.go]: lol my life

**Comment by cpuguy83** [comment]
LGTM

**Comment by calavera** [comment]
LGTM

**Comment by cpuguy83** [comment]
@jfrazelle Found the right incantation to get the Timeout info... it has a nice gift-wrapping:

```go
		if urlErr, ok := err.(*url.Error); ok {
			if opErr, ok := urlErr.Err.(*net.OpError); ok {
				isTimeout = opErr.Timeout()
			}
		}
```

**Comment by jfrazelle** [comment]
oooo pR it!

**Pull request merged**